### DISCUSSION TITLE
Need help implementing KGE for calibration

### DISCUSSION THREAD

**Discussion by Rajiv-0** [neuralhydrology/neuralhydrology#120] [Q&A] [answered]
Similar to MaskedNSELoss, I wrote few lines of codes for MaskedKGELoss. But when I tried to calibrate my model using KGE as loss function, I get unusual values for the loss and average KGE values. Am I missing something here? (Note: The model is calibrated very well using NSE)

**Comment by kratzert** [comment]
Without you sharing your drafted code, how do you think we could help you? Also, I am not sure how you are able to compute the components of the KGE from the samples in the batch.

**Reply by Rajiv-0** [reply]
Sorry for the confusion. I am quite new in neural hydrology so I just need direction on how to proceed with this problem. I added these lines of code in my loss.py file to define the KGE function.
class MaskedKGELoss(BaseLoss):
    def __init__(self, cfg: Config, eps: float = 0.1):
        super(MaskedKGELoss, self).__init__(cfg, prediction_keys=['y_hat'], ground_truth_keys=['y'])
        self.eps = eps

    def _get_loss(self, prediction: Dict[str, TensorType['batch', 'time', 'features']],
                  ground_truth: Dict[str, TensorType['batch', 'time', 'features']], **kwargs) -> TensorType['batch']:
        y = ground_truth['y']
        y_hat = prediction['y_hat']
        my = y - torch.mean(y)
        my_hat = torch.mean(y_hat)
        ym = y - my
        ym_hat = y - my_hat
        epsilon = 10e-5
        r_num = torch.sum(ym * ym_hat)
        std_obs = torch.std(y)
        std_sim = torch.std(y_hat)
        mean_obs = torch.mean(y)
        mean_sim = torch.mean(y_hat)
        y_square_sum = torch.sum(ym * ym)
        y_hat_square_sum = torch.sum(ym_hat * ym_hat)
        r_den = torch.sqrt(y_square_sum * y_hat_square_sum)
        r = torch.mean(r_num/(r_den + epsilon))
        gamma = (std_sim / std_obs)
        beta = mean_sim / mean_obs

        kge_loss = 1 - torch.sqrt((r - 1) ** 2 + (gamma - 1) ** 2 + (beta - 1) ** 2 + self.eps)

        return kge_loss

    @staticmethod
    def _subset_additional_data(additional_data: Dict[str, TensorType['batch', 'time', 'features']],
                                n_target: int) -> Dict[str, TensorType['batch', 'time', 'features']]:
        # No need to subset any additional data
        return additional_data

**Reply by kratzert** [reply]
Are you training your models on a single basin (which you should not!) or on multiple basins at once?.

**Reply by Rajiv-0** [reply]
Thank you for your reply. For this particular test/implementation, I trained on only one basin. However, for all other tests/experiments, I train on a large number of basins at once. I was planning to train on the entire set of catchments once I get a reasonable output using this code.

Is this the correct way to implement the KGE loss, if I want to calibrate my model using KGE? What are the things I need to consider for correct implementation? With this, I am getting a large value of KGE and the losses. But using NSE, even this single basin is trained properly. I feel like I am missing a big picture here. Can you please write some suggestions on how can I deal with this problem? or where I should be looking to make it work?

**Comment by danklotz** [comment]
Hi, 

I am not sure if this will help you or solve the problem, but note that in Deep Learning we *minimize* a loss (that is: we are **not** *maximizing* a performance criterion). If you look at the code for the NH NSE approximation (i.e.: `MaskedNSELoss`) you will see that the `1-` part of the operation is missing. We transformed the NSE into a loss. This is done for exactly that reason.

So if you really wrote the code as you indicate [here](https://github.com/neuralhydrology/neuralhydrology/discussions/120#discussioncomment-6034779) I would suggest to first remove the `1-` in the loss and try again. Currently it seems that the code tries to minimize the KGE. For evaluation NH already has an explicit KGE, so you don't need to worry about that part.

**Reply by Rajiv-0** [reply]
Hi,
Thank you for the suggestion. I tried it too but for some reason it does not seem to be converging. The error as well as KGE keeps getting larger and larger. I am definitely missing something here. :/

**Reply by danklotz** [reply, marked as answer]
Hm... My guess is that some other hyperparameters that you use are off.  I just used your code with my loss-reformulation that I proposed for a single-basin LSTM training and saw that it works. The learning dynamic was quite erratic though, which, e.g., means that you need to use small learning rates and gradient clip carefully. You can also calculate the correlation via the cosine similarity (which I did). I like to think that it makes the training a little bit more stable and the estimation more precise (but haven't checked; anyways, it makes the code simpler).  Can you show me your config, maybe I can see something? 

On a side note, I'd also like to point out  something that @kratzert already hinted at. That is, if you use your loss in a multi-basin setting (which you should if you want to have good models) then you will not optimize for the best KGE per basin. Be careful in that regard.

**Reply by Rajiv-0** [reply]
Hi danklotz. Thank you for your valuable suggestions. I was able to get a good model trained on KGE. I also tried the cosine similarity and it made my code shorter. I did some hyperparameter tuning to get the best sets of parameters for the model and it is more stable now.

**Reply by Rajiv-0** [reply]
Hi danklotz. A quick question. The kge loss here works very well with a single basin  LSTM training. I tried it with many basins and with the use of good parameters sets, it works well. However, now, if I want to apply it to the combination of basins, how should I approach the problem? I suppose I need some sort of modifications in the loss function as the basins have high variations in mean discharges and other characteristics.

**Reply by danklotz** [reply]
Hi @Rajiv-0 

Good to hear that it works well. The issue you are describing now is what I've been warning you about the whole. If you learn with your version of the KGE on multiple basins you will **not** find an optimum for the mean KGE of each basin, but tread the problem *as if** you would just model a single basin. For the mean and the variance terms in the KGE you could use the same trick that we use for the NSE and estimate the observed ones from the training data. However, I see no (straightforward) way to do the same for the Pearson correlation.

**Reply by Rajiv-0** [reply]
Hi @danklotz . Thank you for the response. I will keep working on it and see if something comes up. But still I am surprised to see how well the model works even on a single basin problem using both NSE and KGE. Thank you so much for your time and replies.

**Reply by danklotz** [reply]
Ok, good luck @Rajiv-0!

Yeah, it depends a bit on the basin, but a single basin LSTM is not bad per-se it just has problems with specific processes (such as floods and droughts)

**Reply by sigridjb** [reply]
Hi, I want to use basin-averaged KGE as loss function in a multi-basin calibration, and I am not able to figure out how to do it (specifically, how to extract the per-basin series of observations and simulations to compute the terms in KGE).
According to danklotz' answer there are no straightforward solution, but since two years has passed, I thought is was worth asking again:
 
How can basin-averaged KGE be implemented in loss.py? Or is there maybe a reason why basin-averaged KGE is not implemented in neuralhydrology?

Thanks

**Reply by rajeev-ae** [reply]
Hi @sigridjb . It seems like KGE is more sensitive to per catchments statistics. So when we try to calibrate multiple catchments on KGE, it becomes unstable and breaks. I tried to implement it but was unable to make it work in calibration. However, I tried it in finetuning and the results were interesting. Maybe you can give it a try as well.

**Reply by sigridjb** [reply]
Thank you for the helpful reply. Since you experienced it being unstable and break, does that mean you were able to implement a basin-averaged KGE in loss.py? This is where I struggle (i.e. how to extract per-basin statistics from y_hat, and cov(y, y_hat) per basin). 
Good idea to use it for finetuning, thanks.

**Comment by rodaguayo** [comment]
Hi @Rajiv-0 , can you share the `_get_loss` ? I was trying to do the same
